File: app/lib/models/support_ticket_model.dart
class SupportTicket {
  final String? id;
  final String userName;
  final String userEmail;
  final String? userPhone;
  final String communityName;
  final String subject;
  final String category;
  final String description;
  final String? reproductionSteps;
  final List<String>? attachments;
  final String deviceType;
  final String appVersion;
  final DateTime createdAt;
  final String status;
  final String? assignedTo;
  final DateTime? updatedAt;
  final String? response;
  final DateTime? respondedAt;

  SupportTicket({
    this.id,
    required this.userName,
    required this.userEmail,
    this.userPhone,
    required this.communityName,
    required this.subject,
    required this.category,
    required this.description,
    this.reproductionSteps,
    this.attachments,
    required this.deviceType,
    required this.appVersion,
    required this.createdAt,
    this.status = 'pending',
    this.assignedTo,
    this.updatedAt,
    this.response,
    this.respondedAt,
  });

  factory SupportTicket.fromJson(Map<String, dynamic> json) {
    return SupportTicket(
      id: json['_id'],
      userName: json['userName'],
      userEmail: json['userEmail'],
      userPhone: json['userPhone'],
      communityName: json['communityName'],
      subject: json['subject'],
      category: json['category'],
      description: json['description'],
      reproductionSteps: json['reproductionSteps'],
      attachments: json['attachments'] != null ? List<String>.from(json['attachments']) : null,
      deviceType: json['deviceType'],
      appVersion: json['appVersion'],
      createdAt: DateTime.parse(json['createdAt']),
      status: json['status'] ?? 'pending',
      assignedTo: json['assignedTo'],
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      response: json['response'],
      respondedAt: json['respondedAt'] != null ? DateTime.parse(json['respondedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'userName': userName,
      'userEmail': userEmail,
      'userPhone': userPhone,
      'communityName': communityName,
      'subject': subject,
      'category': category,
      'description': description,
      'reproductionSteps': reproductionSteps,
      'attachments': attachments,
      'deviceType': deviceType,
      'appVersion': appVersion,
      'createdAt': createdAt.toIso8601String(),
      'status': status,
      'assignedTo': assignedTo,
      'updatedAt': updatedAt?.toIso8601String(),
      'response': response,
      'respondedAt': respondedAt?.toIso8601String(),
    };
  }

  SupportTicket copyWith({
    String? id,
    String? userName,
    String? userEmail,
    String? userPhone,
    String? communityName,
    String? subject,
    String? category,
    String? description,
    String? reproductionSteps,
    List<String>? attachments,
    String? deviceType,
    String? appVersion,
    DateTime? createdAt,
    String? status,
    String? assignedTo,
    DateTime? updatedAt,
    String? response,
    DateTime? respondedAt,
  }) {
    return SupportTicket(
      id: id ?? this.id,
      userName: userName ?? this.userName,
      userEmail: userEmail ?? this.userEmail,
      userPhone: userPhone ?? this.userPhone,
      communityName: communityName ?? this.communityName,
      subject: subject ?? this.subject,
      category: category ?? this.category,
      description: description ?? this.description,
      reproductionSteps: reproductionSteps ?? this.reproductionSteps,
      attachments: attachments ?? this.attachments,
      deviceType: deviceType ?? this.deviceType,
      appVersion: appVersion ?? this.appVersion,
      createdAt: createdAt ?? this.createdAt,
      status: status ?? this.status,
      assignedTo: assignedTo ?? this.assignedTo,
      updatedAt: updatedAt ?? this.updatedAt,
      response: response ?? this.response,
      respondedAt: respondedAt ?? this.respondedAt,
    );
  }
}

enum SupportCategory {
  technical('Técnico', '🔧'),
  bug('Error/Bug', '🐛'),
  feature('Nueva Funcionalidad', '✨'),
  account('Cuenta/Acceso', '👤'),
  billing('Facturación', '💰'),
  other('Otro', '❓');

  const SupportCategory(this.label, this.icon);
  final String label;
  final String icon;
}

File: api/src/models/SupportTicket.model.js
const mongoose = require("mongoose");

const supportTicketSchema = new mongoose.Schema(
  {
    userName: {
      type: String,
      required: [true, "El nombre del usuario es requerido"],
      trim: true,
    },
    userEmail: {
      type: String,
      required: [true, "El email del usuario es requerido"],
      trim: true,
      lowercase: true,
    },
    userPhone: {
      type: String,
      trim: true,
    },
    communityName: {
      type: String,
      required: [true, "El nombre de la comunidad es requerido"],
      trim: true,
    },
    subject: {
      type: String,
      required: [true, "El asunto es requerido"],
      trim: true,
      minlength: [10, "El asunto debe tener al menos 10 caracteres"],
      maxlength: [200, "El asunto no puede exceder 200 caracteres"],
    },
    category: {
      type: String,
      required: [true, "La categoría es requerida"],
      enum: {
        values: ["technical", "bug", "feature", "account", "billing", "other"],
        message: "Categoría no válida",
      },
    },
    description: {
      type: String,
      required: [true, "La descripción es requerida"],
      trim: true,
      minlength: [20, "La descripción debe tener al menos 20 caracteres"],
      maxlength: [2000, "La descripción no puede exceder 2000 caracteres"],
    },
    reproductionSteps: {
      type: String,
      trim: true,
      maxlength: [1000, "Los pasos no pueden exceder 1000 caracteres"],
    },
    attachments: {
      type: [String], // Array de URLs de las imágenes
      validate: {
        validator: function (v) {
          return v.length <= 3; // Máximo 3 imágenes
        },
        message: "No se pueden adjuntar más de 3 imágenes",
      },
      default: [],
    },
    deviceType: {
      type: String,
      required: [true, "El tipo de dispositivo es requerido"],
      enum: {
        values: ["Android", "iOS"],
        message: "Tipo de dispositivo no válido",
      },
    },
    appVersion: {
      type: String,
      required: [true, "La versión de la app es requerida"],
      trim: true,
    },
    status: {
      type: String,
      enum: {
        values: ["pending", "in_progress", "resolved", "closed"],
        message: "Estado no válido",
      },
      default: "pending",
    },
    assignedTo: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
    },
    response: {
      type: String,
      trim: true,
      maxlength: [2000, "La respuesta no puede exceder 2000 caracteres"],
    },
    respondedAt: {
      type: Date,
    },
  },
  {
    timestamps: true, // Agrega createdAt y updatedAt automáticamente
  }
);

// Índices para mejorar el rendimiento de las consultas
supportTicketSchema.index({ userEmail: 1, createdAt: -1 });
supportTicketSchema.index({ status: 1, createdAt: -1 });
supportTicketSchema.index({ category: 1, createdAt: -1 });

// Método para obtener tickets del usuario
supportTicketSchema.statics.getUserTickets = function (userEmail) {
  return this.find({ userEmail }).sort({ createdAt: -1 });
};

// Método para obtener tickets por estado
supportTicketSchema.statics.getTicketsByStatus = function (status) {
  return this.find({ status }).sort({ createdAt: -1 });
};

// Método para asignar ticket
supportTicketSchema.methods.assignTo = function (userId) {
  this.assignedTo = userId;
  this.status = "in_progress";
  return this.save();
};

// Método para responder ticket
supportTicketSchema.methods.respond = function (response) {
  this.response = response;
  this.respondedAt = new Date();
  this.status = "resolved";
  return this.save();
};

// Método para cerrar ticket
supportTicketSchema.methods.close = function () {
  this.status = "closed";
  return this.save();
};

// Middleware pre-save para validaciones adicionales
supportTicketSchema.pre("save", function (next) {
  // Si hay pasos de reproducción, validar que no esté vacío
  if (this.reproductionSteps && this.reproductionSteps.trim().length === 0) {
    this.reproductionSteps = undefined;
  }

  // Si hay adjuntos, validar que no estén vacíos
  if (this.attachments && this.attachments.length > 0) {
    this.attachments = this.attachments.filter(
      (url) => url && url.trim().length > 0
    );
  }

  next();
});

module.exports = mongoose.model("SupportTicket", supportTicketSchema);
